draft-yasskin-httpbis-origin-signed-exchanges-impl.md

```diff
@@ -369,86 +428,93 @@ Potentially-valid results include:
-1. If `integrity` names a header field other than `MI-Draft2`
-   ({{!I-D.thomson-http-mice}}) or this header field is not present in
-   `exchange`'s response headers or which the client cannot use to check the
-   integrity of `payload` (for example, the header field is new and hasn't been
-   implemented yet), then return "invalid". Clients MUST be able to check the
-   integrity of `payload` using the `MI-Draft2` header field, which is defined
-   equivalently to the `MI` header from draft 02 of {{!I-D.thomson-http-mice}}.
+1. If `integrity` names a header field that is not present in `exchange`'s
+   response headers or which the client cannot use to check the integrity of
+   `payload` (for example, the header field is new and hasn't been implemented
+   yet), then return "invalid". If the selected header field provides integrity
+   guarantees weaker than SHA-256, return "invalid". If validating integrity
+   using the selected header field requires the client to process records larger
+   than TBD bytes, return "invalid". Clients MUST implement at least the `MI`
+   ({{!I-D.thomson-http-mice}}) header field with its `mi-sha256` content
+   encoding.
 1. Set `publicKey` and `signing-alg` depending on which key fields are present:
-   1. Assert: `certUrl` is present.
+   1. If `cert-url` is present:
       1. Let `certificate-chain` be the result of loading the certificate chain
-         at `certUrl` passing the `forceFetch` flag ({{cert-chain-format}}). If
+         at `cert-url` passing the `forceFetch` flag ({{cert-chain-format}}). If
          this returns "invalid", return "invalid".
       1. Let `main-certificate` be the first certificate in `certificate-chain`.
       1. Set `publicKey` to `main-certificate`'s public key.
-      1. If `publicKey` is not a 2048-bit RSA public key, return "invalid".
-      1. The client MUST define a partial function from public key types to
-         signing algorithms, and this function must at the minimum include the
-         following mappings:
-
-         RSA, 2048 bits:
-         : rsa_pss_rsae_sha256 or rsa_pss_pss_sha256, as defined in Section
-           4.2.3 of {{!I-D.ietf-tls-tls13}}, depending on which of the
-           rsaEncryption OID or RSASSA-PSS OID {{!RFC8017}} is used.
-
-         Set `signing-alg` to the result of applying this function to the type of
-         `main-certificate`'s public key. If the function is undefined on this
-         input, return "invalid".
+      1. If `publicKey` is an RSA key, return "invalid".
+      1. If `publicKey` is a key using the secp256r1 elliptic curve, set
+         `signing-alg` to ecdsa_secp256r1_sha256 as defined in Section 4.2.3 of
+         {{!I-D.ietf-tls-tls13}}.
+      1. Otherwise, either return "invalid" or set `signing-alg` to a non-legacy
+         signing algorithm defined by TLS 1.3 or later
+         ({{!I-D.ietf-tls-tls13}}). This choice MUST depend only on
+         `publicKey`'s type and not on any other context.
```

Comment:
1. Does this mean the choice cannot depend on the signature_algorithms_cert extension?
2. Can the choice be "always return invalid"? (That appears to be what [chromium](https://cs.chromium.org/chromium/src/content/browser/web_package/signed_exchange_signature_verifier.cc?l=183&rcl=b7527b9711e28030e03ad438ba323af7142ce43b) does IIUC.)

(Pardon my naivety!)

Reply:
Out of curiosity, is there a specific "other context" that this language intends to prevent?

Reply:
For example, lots of signature formats, like [JWS](https://tools.ietf.org/html/rfc7515#section-5.2) include an algorithm name in the signature in addition to a public key identifier. This allows some silly things like https://www.sjoerdlangkemper.nl/2016/09/28/attacking-jwt-authentication/.